x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportPutTransformAction.java

```diff
@@ -92,61 +92,60 @@ public TransportPutTransformAction(
-                .setVersion(Version.CURRENT);
-
-            String transformId = config.getId();
-            // quick check whether a transform has already been created under that name
-            if (PersistentTasksCustomMetadata.getTaskWithId(clusterState, transformId) != null) {
-                listener.onFailure(
-                    new ResourceAlreadyExistsException(
-                        TransformMessages.getMessage(TransformMessages.REST_PUT_TRANSFORM_EXISTS, transformId)
-                    )
-                );
-                return;
-            }
-
-            // <3> Create the transform
-            ActionListener<ValidateTransformAction.Response> validateTransformListener = ActionListener.wrap(
-                validationResponse -> putTransform(request, listener),
-                listener::onFailure
+        String transformId = config.getId();
+        // quick check whether a transform has already been created under that name
+        if (PersistentTasksCustomMetadata.getTaskWithId(clusterState, transformId) != null) {
+            listener.onFailure(
+                new ResourceAlreadyExistsException(TransformMessages.getMessage(TransformMessages.REST_PUT_TRANSFORM_EXISTS, transformId))
             );
+            return;
+        }
 
-            // <2> Validate source and destination indices
-            ActionListener<Void> checkPrivilegesListener = ActionListener.wrap(
-                aVoid -> client.execute(
-                    ValidateTransformAction.INSTANCE,
-                    new ValidateTransformAction.Request(config, request.isDeferValidation(), request.timeout()),
-                    validateTransformListener
-                ),
-                listener::onFailure
-            );
+        // <3> Create the transform
+        ActionListener<ValidateTransformAction.Response> validateTransformListener = ActionListener.wrap(
+            validationResponse -> putTransform(request, listener),
+            listener::onFailure
+        );
 
-            // <1> Early check to verify that the user can create the destination index and can read from the source
-            if (XPackSettings.SECURITY_ENABLED.get(settings) && request.isDeferValidation() == false) {
-                TransformPrivilegeChecker.checkPrivileges(
-                    "create",
-                    securityContext,
-                    indexNameExpressionResolver,
-                    clusterState,
-                    client,
-                    config,
-                    true,
-                    checkPrivilegesListener
-                );
-            } else { // No security enabled, just move on
-                checkPrivilegesListener.onResponse(null);
-            }
-        });
+        // <2> Validate source and destination indices
+        ActionListener<Void> checkPrivilegesListener = ActionListener.wrap(
+            aVoid -> ClientHelper.executeWithHeadersAsync(
+                config.getHeaders(),
```

Comment:
I find it surprising that this works. `config.setHeaders` was called inside a block that switched to secondary headers. So I would have thought that here we end up called `ValidateTransformAction` using the secondary headers, which should fail if those secondary headers only allow index access and not transform administration.

I can see you have a test that seems to test that this functionality works, but please can you do some extra due diligence on whether this is really working as expected, perhaps by running the test locally with extra debug, or by manually testing it with `gradle run`.

If this extra testing shows that the existing test is only succeeding by some fluke then I think the way to fix it would be to put this call back to how it was before, and instead switch to secondary headers (using `useSecondaryAuthIfAvailable`) within the subsections of `TransportValidateTransformAction` that access the source or destination indices.

If the code really does work correctly as you've got it now then I need to adjust my understanding of what `executeWithHeadersAsync` does.

Reply:
> So I would have thought that here we end up calling ValidateTransformAction using the secondary headers, which should fail if those secondary headers only allow index access and not transform administration.

Yes, that was indeed the case and the scenario from Quynh I reproduced.

> I can see you have a test that seems to test that this functionality works, but please can you do some extra due diligence on whether this is really working as expected, perhaps by running the test locally with extra debug, or by manually testing it with gradle run.

So, I did a little back and forth on this PR and the version you reviewed probably had the test (my new test written exactly for this case) failing.

The reason for this was I noticed Ben's work (https://github.com/elastic/elasticsearch/pull/86757) solving the issue https://github.com/elastic/elasticsearch/issues/86731 (Prefer secondary authentication headers when creating/updating transforms) and I wanted to keep his tests passing as well.

I ended up with the following (this PR is now updated):
- pass **all** the security headers to `_validate` call
- only store **secondary** headers in the transform config.

This way `_validate` call can be executed even if the secondary headers does not have `transform_admin`.
And at the same time we only store secondary header in the config in order to use it for data access later on (e.g: creating destination index).